File: hackerrank/problem solving/algorithms/2 - implementation/34 - cut the sticks/oo/cut_the_sticks_oo.js
// Source: https://www.hackerrank.com/challenges/cut-the-sticks/problem?isFullScreen=true

'use strict';

process.stdin.resume();
process.stdin.setEncoding('utf-8');

let inputString = '';
let currentLine = 0;

process.stdin.on('data', function(inputStdin) {
    inputString += inputStdin;
});

process.stdin.on('end', function() {
    inputString = inputString.split('\n');
    main();
});

function readLine() {
    return inputString[currentLine++];
}


function main() {
    const _SIZE = +readLine();
    const ARRAY = readAnArray();
    ARRAY.sort((a, b) => a - b);

    const OBJ = new CutTheSticks(ARRAY);
    const RESULT = OBJ.getRemainingSticksLengthArray();
    printArray(RESULT);
}

    function readAnArray() {
        return readLine().split(" ").map(Number);
    }

    function printArray(array) {
        array.forEach((element, _) => console.log(element));
    }

    class CutTheSticks {
        #sticksLengthArray;
        #sizeSticksLengthArray;
        #remainingSticksLengthArray;

        constructor(array) {
            this.#sticksLengthArray = array;
            this.#sizeSticksLengthArray = array.length;
            this.#initializeRemainingSticksLengthArray();
            this.#cutTheSticks();
        }

            #initializeRemainingSticksLengthArray() {
                this.#remainingSticksLengthArray = [this.#sizeSticksLengthArray];
            }

            #cutTheSticks() {
                for (let i = 0; i < this.#sizeSticksLengthArray; ) {
                    const SHORTEST_STICK_LENGTH = this.#sticksLengthArray[i];

                    i = this.#findFirstElementIndexDifferentOfFirstCurrentShortestStickLengthIndex(i);
                    const CURRENT_ITERATION_SIZE = this.#sizeSticksLengthArray - i;
                    if (i === this.#sizeSticksLengthArray)
                        break;

                    this.#remainingSticksLengthArray.push(CURRENT_ITERATION_SIZE);
                    this.#decreaseSticksLengthsHigherThanCurrentShortestStickLength(i, SHORTEST_STICK_LENGTH);
                }
            }

                #findFirstElementIndexDifferentOfFirstCurrentShortestStickLengthIndex(index) {
                    const SHORTEST_STICK_LENGTH = this.#sticksLengthArray[index];
                    while (index < this.#sizeSticksLengthArray && this.#sticksLengthArray[index] === SHORTEST_STICK_LENGTH)
                        index++;
                    return index;
                }

                #decreaseSticksLengthsHigherThanCurrentShortestStickLength(index, shortestStickLength) {
                    for (let i = index; i < this.#sizeSticksLengthArray; i++)
                        this.#sticksLengthArray[i] -= shortestStickLength;
                }

        getRemainingSticksLengthArray() {
            return this.#remainingSticksLengthArray;
        }
    }

File: hackerrank/problem solving/algorithms/2 - implementation/34 - cut the sticks/oo/cut_the_sticks_oo.py
# Source: https://www.hackerrank.com/challenges/cut-the-sticks/problem?isFullScreen=true

def main():

    _SIZE = int(input())
    ARRAY = readAnArray()
    ARRAY.sort()

    OBJ = CutTheSticks(ARRAY)
    RESULT = OBJ.getRemainingSticksLengthArray()
    printArray(RESULT)


def readAnArray():

    return list(map(int, input().split()))


def printArray(array):

    print(*array, sep="\n")


class CutTheSticks:

    def __init__(self, array):

        self.__sticksLengthArray = array
        self.__SIZE_STICKS_LENGTH_ARRAY = len(array)
        self.__remainingSticksLengthArray = self.__initializeRemainingSticksLengthArray()
        self.__cutTheSticks()


    def __initializeRemainingSticksLengthArray(self):

        return [self.__SIZE_STICKS_LENGTH_ARRAY]


    def __cutTheSticks(self):

        i = 0
        while i < self.__SIZE_STICKS_LENGTH_ARRAY:
            SHORTEST_STICK_LENGTH = self.__sticksLengthArray[i]

            i = self.__findFirstElementIndexDifferentOfFirstCurrentShortestStickLengthIndex(i)
            CURRENT_ITERATION_SIZE = self.__SIZE_STICKS_LENGTH_ARRAY - i
            if i == self.__SIZE_STICKS_LENGTH_ARRAY:
                break

            self.__remainingSticksLengthArray.append(CURRENT_ITERATION_SIZE)
            self.__decreaseSticksLengthsHigherThanCurrentShortestStickLength(i, SHORTEST_STICK_LENGTH)


    def __findFirstElementIndexDifferentOfFirstCurrentShortestStickLengthIndex(self, index):

        SHORTEST_STICK_LENGTH = self.__sticksLengthArray[index]
        while index < self.__SIZE_STICKS_LENGTH_ARRAY and self.__sticksLengthArray[index] == SHORTEST_STICK_LENGTH:
            index += 1
        return index


    def __decreaseSticksLengthsHigherThanCurrentShortestStickLength(self, index, shortestStickLength):

        for i in range(index, self.__SIZE_STICKS_LENGTH_ARRAY):
            self.__sticksLengthArray[i] -= shortestStickLength


    def getRemainingSticksLengthArray(self):

        return self.__remainingSticksLengthArray


if __name__ == "__main__":
    main()
